packages/patrol_cli/lib/src/coverage/coverage_tool.dart

```diff
@@ -0,0 +1,213 @@
+      connectionDetails.webSocketUri.toString(),
+    );
+    _disposeScope.addDispose(serviceClient.dispose);
+
+    await serviceClient.streamListen('Extension');
+    final completer = Completer<int>()..disposedBy(_disposeScope, 0);
+    serviceClient.onExtensionEvent.listen((event) async {
+      if (event.extensionKind == 'testCount') {
+        completer.complete(event.extensionData!.data['testCount'] as int);
+      }
+    }).disposedBy(_disposeScope);
+
+    final testCount = await completer.future;
+    await serviceClient.dispose();
+
+    return testCount;
+  }
+
+  Future<Map<String, coverage.HitMap>> _collectFromVM({
+    required String flutterPackageName,
+    required VMConnectionDetails connectionDetails,
+  }) async {
+    final result = <String, coverage.HitMap>{};
+    final serviceClient = await vmServiceConnectUri(
+      connectionDetails.webSocketUri.toString(),
+    );
+    _disposeScope.addDispose(serviceClient.dispose);
+    await serviceClient.streamListen('Extension');
+    final event = await serviceClient.onExtensionEvent
+        .where((event) => event.extensionKind == 'waitForCoverageCollection')
+        .first;
+    result.merge(
+      await _collectAndMarkTestCompleted(
+        connectionDetails: connectionDetails,
+        packageName: flutterPackageName,
+        mainIsolateId: event.extensionData!.data['mainIsolateId'] as String,
+      ),
+    );
+    await serviceClient.dispose();
+
+    return result;
+  }
+
+  Future<Map<String, coverage.HitMap>> _collectAndMarkTestCompleted({
+    required VMConnectionDetails connectionDetails,
+    required String packageName,
+    required String mainIsolateId,
+  }) async {
+    final data = await coverage.collect(
+      connectionDetails.uri,
+      false,
+      false,
+      false,
+      {packageName},
+    );
+
+    final socket =
+        await WebSocket.connect(connectionDetails.webSocketUri.toString())
+          ..add(
+            jsonEncode(
+              {
+                'jsonrpc': '2.0',
+                'id': 21,
+                'method': 'ext.patrol.markTestCompleted',
+                'params': {
+                  'isolateId': mainIsolateId,
+                  'command': 'markTestCompleted',
+                },
+              },
+            ),
+          );
+    await socket.close();
+
+    return coverage.HitMap.parseJson(
+      data['coverage'] as List<Map<String, dynamic>>,
+    );
+  }
+
+  Future<void> _saveReport(String report) async {
+    final coverageDirectory = _fs.directory('coverage');
+
+    if (!coverageDirectory.existsSync()) {
+      await coverageDirectory.create();
+    }
+
+    await coverageDirectory.childFile('patrol_lcov.info').writeAsString(report);
+  }
+}
+
+extension<T> on Completer<T> {
+  void disposedBy(DisposeScope disposeScope, T disposeValue) {
+    disposeScope.addDispose(() {
+      if (!isCompleted) {
+        complete(disposeValue);
+      }
+    });
+  }
+}
```

Comment:
This seems like a good addition to [`package:dispose_scope`](https://pub.dev/packages/dispose_scope) by @RobertOdrowaz.